File: elixir/collatz-conjecture/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(1), do: 0
  def calc(input) when input > 1 and is_integer(input) do
    cond do
      rem(input, 2) === 0 -> 1 + calc(div(input, 2))
      true -> 1 + calc(3 * input + 1)
    end
  end
end

File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz n 
    | n > 0 = Just (collatzConj 0 n)
    | otherwise = Nothing
    where
        collatzConj steps 1 = steps
        collatzConj steps n = collatzConj (1 + steps) (if even n then n `div` 2 else 3 * n + 1)
